bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -4011,7 +4011,24 @@ void ReplicaImp::executeRequestsAndSendResponses(PrePrepareMsg *ppMsg,
       free(req.outReply);
       send(replyMsg.get(), req.clientId);
     }
-    clientsManager->removePendingRequestOfClient(req.clientId);
+  }
+}
+
+void ReplicaImp::tryToRemovePendingRequestsForSeqNum(SeqNum seqNum) {
+  SCOPED_MDC_SEQ_NUM(std::to_string(seqNum));
+  SeqNumInfo &seqNumInfo = mainLog->get(seqNum);
+  PrePrepareMsg *prePrepareMsg = seqNumInfo.getPrePrepareMsg();
+  if (prePrepareMsg == nullptr) return;
+  LOG_INFO(GL, "clear pending requests" << KVLOG(seqNum));
+
+  RequestsIterator reqIter(prePrepareMsg);
+  char *requestBody = nullptr;
+  while (reqIter.getAndGoToNext(requestBody)) {
+    ClientRequestMsg req((ClientRequestMsgHeader *)requestBody);
+    if (!clientsManager->isValidClient(req.clientProxyId())) continue;
```

Comment:
Same as above, might be we should warn that something wrong has happened here.

Reply:
This a valid situation, for example, with NOOP commnads